system_zero/src/permutation_unit.rs

```diff
@@ -2,36 +2,117 @@ use plonky2::field::extension_field::{Extendable, FieldExtension};
 use plonky2::field::packed_field::PackedField;
 use plonky2::hash::hash_types::RichField;
 use plonky2::hash::hashing::SPONGE_WIDTH;
+use plonky2::hash::poseidon::{HALF_N_FULL_ROUNDS, N_PARTIAL_ROUNDS};
 use plonky2::plonk::circuit_builder::CircuitBuilder;
 use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 use starky::vars::StarkEvaluationTargets;
 use starky::vars::StarkEvaluationVars;
 
-use crate::column_layout::{col_permutation_input, col_permutation_output, NUM_COLUMNS};
+use crate::column_layout::{
+    col_permutation_full_first_after_mds, col_permutation_full_first_mid_sbox,
+    col_permutation_full_second_after_mds, col_permutation_full_second_mid_sbox,
+    col_permutation_input, col_permutation_partial_after_sbox, col_permutation_partial_mid_sbox,
+    NUM_COLUMNS,
+};
 use crate::public_input_layout::NUM_PUBLIC_INPUTS;
 use crate::system_zero::SystemZero;
 
+fn constant_layer<F, FE, P, const D2: usize>(
+    mut state: [P; SPONGE_WIDTH],
+    round: usize,
+) -> [P; SPONGE_WIDTH]
+where
+    F: RichField,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    // One day I might actually vectorize this, but today is not that day.
```

Comment:
This should work right?

```rust
for i in 0..SPONGE_WIDTH {
    state[i] += P::from(FE::from_canonical_u64(ALL_ROUND_CONSTANTS[i + SPONGE_WIDTH * round]));
}
```

Reply:
I didn't do this to avoid copying the Poseidon logic into the permutation unit (as much as possible, anyway). I'd rather make the Poseidon function work with packed fields, which will be easy, but is a different PR.